e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -233,24 +214,107 @@ func (pf *PortForwarder) start() error {
-					return
-				}
-				// Watch automatically retries, so this shouldn't happen unless
-				// there's a terminal error.
-				pf.logger.Infof("Watch exited, restarting port-forward for %s/%s: %v", pf.ns, pf.deployment, err)
-			}
-			if err := pf.portForwardToDeployment(true); err != nil {
-				pf.logger.Info(err)
-				// this is a potentially fatal error since we failed to restart the port
-				// forward. we log the error and return in case the test is already done
-				// with the port forward. the test will otherwise error the next time it
-				// tries to use the local port.
+			if err := pf.waitAndRestart(); err != nil {
+				pf.logger.Infof("PortForwarder finished: %v", err)
 				return
 			}
 		}
 	}()
 	return nil
 }
+
+// waitAndRestart will block until seeing an indicator that the port-forward is
+// no longer healthy. If that occurs, it will restart the port-forward
+func (pf *PortForwarder) waitAndRestart() error {
+	err := pf.waitForFailure()
+	if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+		return errors.Wrap(err, fmt.Sprintf("exiting - PortForwarder for pod %s/%s received done signal", pf.ns, pf.pod))
+	}
+	pf.logger.Infof("received signal for port-forward for %s/%s. err: %v", pf.ns, pf.pod, err)
+	if err := pf.portForwardToDeployment(true); err != nil {
+		// this is a potentially fatal error since we failed to restart the port
+		// forward. we log the error and return in case the test is already done
+		// with the port forward. the test will otherwise error the next time it
+		// tries to use the local port.
+		return err
+	}
+	return nil
+}
+
+// waitForFailure waits for any failure condition which indicates the port-forward
+// must be restarted. The port-forward must be restarted if:
+// - the subprocess is killed
+// - the pod is restarted
+func (pf *PortForwarder) waitForFailure() error {
+	ctx, cancel := context.WithCancel(pf.ctx)
+	cmdChan := wrapWithChannel(func() error {
+		return pf.waitForCmdFailure(ctx)
+	})
+	podChan := wrapWithChannel(func() error {
+		return pf.waitForPodFailure(ctx)
+	})
+	defer func() {
+		// before we return, cancel and wait for goroutines to exit
+		cancel()
+		<-cmdChan
+		<-podChan
+	}()
+	// wait for any of the channels to return. a happy little race
+	select {
+	case err := <-cmdChan:
+		return err
+	case err := <-podChan:
+		return err
+	}
+}
+
+// waitForCmdFailure waits for the port-forward subprocess to be killed until ctx is closed.
+// If ctx is closed, the subprocess will be killed before returning.
+// An error is returned to trigger port-forward restart.
+func (pf *PortForwarder) waitForCmdFailure(ctx context.Context) error {
+	// cmd.Wait does not accept a context, so create a wrapper that cancels with the ctx
+	cmdWaitChan := wrapWithChannel(pf.cmd.Wait)
+	var chanErr error
+	select {
+	case <-ctx.Done():
+		// stop waiting for the command to exit, so we can kill it
+		chanErr = ctx.Err()
+	case err := <-cmdWaitChan:
+		chanErr = err
```

Comment:
I think if the cmd exits, you can just return here. You don't need to kill it or wait for it to exit any more.

Reply:
fair point, done